### ISSUE TITLE
libcurl HttpClient incompatibility with HTTP2 protocol

### ISSUE THREAD

**Issue by JustArchi** [dotnet/corefx#27730]
Recently I started running into issues like this one:

```csharp
OnUnobservedTaskException() System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (An error occurred while sending the request.) ---> System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.Net.Http.CurlException: Error in the HTTP2 framing layer
   at System.Net.Http.CurlHandler.ThrowIfCURLEError(CURLcode error)
   at System.Net.Http.CurlHandler.MultiAgent.FinishRequest(StrongToWeakReference`1 easyWrapper, CURLcode messageResult)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Discord.Net.Rest.DefaultRestClient.SendInternalAsync(HttpRequestMessage request, CancellationToken cancelToken, Boolean headerOnly)
   at Discord.Net.Rest.DefaultRestClient.SendAsync(String method, String endpoint, CancellationToken cancelToken, Boolean headerOnly, String reason)
   at Discord.Net.Queue.RestRequest.SendAsync()
   at Discord.Net.Queue.RequestBucket.SendAsync(RestRequest request)
   at Discord.Net.Queue.RequestQueue.SendAsync(RestRequest request)
   at Discord.API.DiscordRestApiClient.SendInternalAsync(String method, String endpoint, RestRequest request)
   at Discord.API.DiscordRestApiClient.SendAsync[TResponse](String method, String endpoint, String bucketId, ClientBucketType clientBucket, RequestOptions options)
   at Discord.API.DiscordSocketApiClient.GetGatewayAsync(RequestOptions options)
   at Discord.API.DiscordSocketApiClient.ConnectInternalAsync()
   at Discord.API.DiscordSocketApiClient.ConnectInternalAsync()
   at Discord.API.DiscordSocketApiClient.ConnectAsync()
   at Discord.WebSocket.DiscordSocketClient.OnConnectingAsync()
   at Discord.WebSocket.DiscordSocketClient.OnConnectingAsync()
   at Discord.ConnectionManager.ConnectAsync(CancellationTokenSource reconnectCancelToken)
   at Discord.ConnectionManager.<>c__DisplayClass28_0.<<StartAsync>b__0>d.MoveNext()
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.Net.Http.CurlException: Error in the HTTP2 framing layer
   at System.Net.Http.CurlHandler.ThrowIfCURLEError(CURLcode error)
   at System.Net.Http.CurlHandler.MultiAgent.FinishRequest(StrongToWeakReference`1 easyWrapper, CURLcode messageResult)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Discord.Net.Rest.DefaultRestClient.SendInternalAsync(HttpRequestMessage request, CancellationToken cancelToken, Boolean headerOnly)
   at Discord.Net.Rest.DefaultRestClient.SendAsync(String method, String endpoint, CancellationToken cancelToken, Boolean headerOnly, String reason)
   at Discord.Net.Queue.RestRequest.SendAsync()
   at Discord.Net.Queue.RequestBucket.SendAsync(RestRequest request)
   at Discord.Net.Queue.RequestQueue.SendAsync(RestRequest request)
   at Discord.API.DiscordRestApiClient.SendInternalAsync(String method, String endpoint, RestRequest request)
   at Discord.API.DiscordRestApiClient.SendAsync[TResponse](String method, String endpoint, String bucketId, ClientBucketType clientBucket, RequestOptions options)
   at Discord.API.DiscordSocketApiClient.GetGatewayAsync(RequestOptions options)
   at Discord.API.DiscordSocketApiClient.ConnectInternalAsync()
   at Discord.API.DiscordSocketApiClient.ConnectInternalAsync()
   at Discord.API.DiscordSocketApiClient.ConnectAsync()
   at Discord.WebSocket.DiscordSocketClient.OnConnectingAsync()
   at Discord.WebSocket.DiscordSocketClient.OnConnectingAsync()
   at Discord.ConnectionManager.ConnectAsync(CancellationTokenSource reconnectCancelToken)
   at Discord.ConnectionManager.<>c__DisplayClass28_0.<<StartAsync>b__0>d.MoveNext()<---
```

I'm almost 100% sure that this issue is the result of some incompatibility with recent libcurl version. I reproduced it with libcurl3 in version 7.58.0 available in Debian testing repo. Temporarily I downgraded myself back to 7.52.1 available in Debian stable and I can't reproduce the issue anymore, at least for now.

The exception itself isn't exactly in my code but in third-party library I'm using, although there is nothing unusual there.

```csharp
        private async Task<RestResponse> SendInternalAsync(HttpRequestMessage request, CancellationToken cancelToken, bool headerOnly)
        {
            cancelToken = CancellationTokenSource.CreateLinkedTokenSource(_cancelToken, cancelToken).Token;
            HttpResponseMessage response = await _client.SendAsync(request, cancelToken).ConfigureAwait(false);
            
            var headers = response.Headers.ToDictionary(x => x.Key, x => x.Value.FirstOrDefault(), StringComparer.OrdinalIgnoreCase);
            var stream = !headerOnly ? await response.Content.ReadAsStreamAsync().ConfigureAwait(false) : null;

            return new RestResponse(response.StatusCode, headers, stream);
        }
```

I'm using one of the recent master builds and I didn't see anything related fixed lately.

```
.NET Command Line Tools (2.1.300-preview2-008293)

Product Information:
 Version:            2.1.300-preview2-008293
 Commit SHA-1 hash:  68922e2a51

Runtime Environment:
 OS Name:     debian
 OS Version:
 OS Platform: Linux
 RID:         debian-x64
 Base Path:   /opt/dotnet/sdk/2.1.300-preview2-008293/

Host (useful for support):
  Version: 2.1.0-preview2-26227-01
  Commit:  86d1f92013

.NET Core SDKs installed:
  2.1.300-preview2-008293 [/opt/dotnet/sdk]

The.NET Core runtimes installed:
  Microsoft.AspNetCore.App 2.1.0-preview2-30171 [/opt/dotnet/shared]
  Microsoft.NETCore.App 2.1.0-preview2-26227-01 [/opt/dotnet/shared]
  Microsoft.AspNetCore.All 2.1.0-preview2-30171 [/opt/dotnet/shared]
```

I'm now making sure that old libcurl version fixes the issue, since I could've identified the root cause wrong, but for now I can't run into this issue anymore with older libcurl version. I'll make sure to let you know in case it's something else, but if my guess was right then this should be fixed in corefx repo, and stacktrace suggests that even if my guess is wrong then there is some incompatibility in net core internals that should be verified.

Also I have no clue under exactly what condition the exception is happening. It's definitely related to HTTP2 connection, but it seems that not all requests going through HTTP2 throw this exception (?)

Thank you in advance for looking into this.

**Comment by davidsh** [comment]
cc: @stephentoub

**Comment by stephentoub** [comment]
Thanks for the report.

> but if my guess was right then this should be fixed in corefx repo, and stacktrace suggests that even if my guess is wrong then there is some incompatibility in net core internals that should be verified

If it has to do with libcurl version, I think it's unlikely that it's a bug in corefx.  corefx doesn't do anything special for HTTP/2; it just asks libcurl to make the request.  What kind of fix are you suggesting corefx employ?

**Comment by JustArchi** [comment]
Well, that depends where the issue exactly is - I didn't do full analysis yet, neither I know corefx internals and how it makes use of libcurl as a backend. If it's purely libcurl issue and corefx code has nothing to do with it then I'd expect that you forwarded the issue where appropriate, just like how I forward it to corefx, you're the boss here 👍

I have one idea which could be beneficial for time being - environment variable that would control default HTTP protocol version to use, such as `DOTNET_SYSTEM_NET_HTTP_DEFAULTPROTOCOL=1.1`, or something simple such as `DOTNET_SYSTEM_NET_HTTP_DEFAULTHTTP2=0`. I realize that it's possible to specify protocol version on http request explicitly, but since `HttpClient` is used in many third-party libraries and utilities, modifying everything globally is problematic. Environment variable or maybe `ServicePointManager` setting would make sense to me.

It'd make even more sense considering HTTP 2.0 is now default (as of .NET Core 2.1). I think HTTP 1.1 was default before (as of .NET Core 2.0), but correct me if I'm wrong. I guess it'd make sense for time being, but it's not directly related to the issue above, just a potential workaround.

**Comment by karelz** [comment]
@JustArchi I think that we would need a repro, ideally directly against libcurl, or at least reasonable network traces, to be able to route it further to libcurl. Currently there's not much else to say beside "sometimes it fails with this error".
Also, I wonder if it may be a sign of not well-behaving HTTP/2 server (depends on when libcurl uses the error) - do you have any signs that it happens only with specific server(s)?

**Comment by JustArchi** [comment]
It's definitely possible that it could be a server-related issue, I can't deny it because the only trace I have right now is from Discord server. On the other hand it's very likely that I have only that one simply because many servers do not support HTTP/2 as of today and it might be the only one I actually establish HTTP/2 connection with.

In any case, I'll try to make some repro, although I'm as puzzled as you are regarding this, if not more. For now the only thing I can be sure about is that the newer version of libcurl brought this issue, so it's entirely possible that the target server didn't work properly for entire time and libcurl upgrade accidentally discovered that problem, or it's some libcurl regression instead that should not happen. I'll let you know if I find out anything interesting, it's good to keep this issue around for now, as other people might stumble upon this too at some point, especially after release of .NET Core 2.1 that will change default protocol version to 2.0.

**Comment by JustArchi** [comment]
Good news - I managed to write reproducible case. Bad news - I managed to reproduce it only with one specific server, but reliably.

And with that reproducible case I confirmed all my initial thoughts:
- I used standard .NET Core `HttpClient` for reproduction, without any third-party tools or libraries.
- Exception reliably pops up when sending exactly 1000th request (no clue why, but it's always 1000th)
- I couldn't reproduce this issue with curl CLI tool in a bash loop (tested past 1000th request)
- I also couldn't reproduce this issue with old (7.52.1 tested) libcurl3 version, I reliably reproduce it with recent (7.58 tested) one.

I attack the reproducible case I could come up with. Please note that it's not really "clean" (we need to send 1000 requests before it pops up), but I can at least say that it's reliable, since I managed to reproduce it 5 times, always when sending 1000th request. Thank you in advance for looking into the issue.

![image](https://user-images.githubusercontent.com/1069029/37036585-653aee8e-2150-11e8-9ee3-05c37fafb70d.png)

[repro.zip](https://github.com/dotnet/corefx/files/1785276/repro.zip)

**Comment by karelz** [comment]
Great! Did you get a chance to run Fiddler on the repro to see if the server communication is bogus vs. if it is internal libcurl hiccup?

I also wonder if you could try a C++ app using libcurl to do the same (1000 requests) - running curl in bash spins up always a new process, so I assume it is not quite the same repro as the .NET app. That would tell us if it is libcurl problem vs. .NET problem.

**Comment by JustArchi** [comment]
Well, my C++ is rather rusty, but I managed to merge two samples together and create C++ libcurl repro as well, so it's not .NET Core issue at all, like you both initially guessed.

```c
#include <stdio.h>
#include <curl/curl.h>
#include <string.h>

int main(void)
{
  CURL *curl;

  curl_global_init(CURL_GLOBAL_DEFAULT);

  curl = curl_easy_init();
  if(curl) {
    curl_easy_setopt(curl, CURLOPT_URL, "https://discordapp.com/api/v6/gateway");
    curl_easy_setopt(curl, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_2_0);

    char errbuf[CURL_ERROR_SIZE];
    curl_easy_setopt(curl, CURLOPT_ERRORBUFFER, errbuf);
    errbuf[0] = 0;

    int count = 1;

    for (count = 1; count < 1500; count++) {
        CURLcode res = curl_easy_perform(curl);
        if (res != CURLE_OK) {
            size_t len = strlen(errbuf);
            fprintf(stderr, "\nlibcurl: (%d) ", res);
            if (len) {
                fprintf(stderr, "%s%s", errbuf, ((errbuf[len - 1] != '\n') ? "\n" : ""));
            } else {
                fprintf(stderr, "%s\n", curl_easy_strerror(res));
            }

            break;
        }

        fprintf(stdout, "OK: #%d\n", count);
    }

    curl_easy_cleanup(curl);
  }

  curl_global_cleanup();

  return 0;
}
```

![image](https://user-images.githubusercontent.com/1069029/37054175-690ac74a-217e-11e8-8e9d-c17a0656cfb3.png)

I'll forward this where appropriate and leave a note, since it's definitely not .NET Core issue at this point. Thanks.

**Comment by bagder** [comment]
I suspect this is [an issue we fixed in libcurl 7.60.0](https://github.com/curl/curl/issues/2416). The 1000 streams is *probably* because a default nginx installation closes the connection then.

**Comment by JustArchi** [comment]
@bagder I reported it in https://github.com/curl/curl/issues/2365 🙂
